### PR TITLE
Adr modular architecture revised

### PR DESCRIPTION
**Pull request by emettely**
**Is your Pull Request request related to another [issue](https://github.com/bbc/react-transcript-editor/issues) in this repository ?**  
-

**Describe what the PR does**    
Updates the ADR for modular architecture.

**State whether the PR is ready for review or whether it needs extra work**    
Ready for Review

**Additional context**    
Following much discussion, we decided to remove NPM packaging for the API. This is the revised modular architecture ADR containing that information and additional information around the discussions that happened in the last two days.

## Definitions of Done
- [ ] Runs locally
- [ ] Runs remotely
- [ ] Test passes
- [ ] Demonstrated
- [ ] Deployed to Cosmos on Test and Live
- [x] Documentation
- [ ] Peer reviewed by: hopefully everyone in the team

### CODE REVIEW

**Comment by emettely** [comment]
For completeness sake:

@emettely 
Re some discussions earlier:
*Why have CI for npm release?*
* ensure clean environment every deployment and clear dependency - every computer that we have has a different local department set up so we want to make sure we have a clear idea on dependency.
* This is also another reason why we want a sandbox generally, to be able to replicate what would happen on Jenkins or a live cosmos machine (both centOS).
* It also provides transparency for other team members to see if test have succeeded or not.
* automate testing
* I’m sure there are other reasons but I really think we should have CI and not do just local development and npm release even if we have a good practice of cleaning up before committing to master. There have been times when locally tests worked, the tests then failed in Jenkins due to some random dependency. This isn’t necessarily because the tests were poorly written or wasn’t exhaustive enough.

*Why is installing dpe-client so slow???*
* remember that the code hasn’t been looked at much, so dependencies need to be cleaned up. This will be a fire and forget thing, because the actual build process will be in Jenkins. All you need to do locally is change the version number and that’s it.

*There is overhead of having the API as an npm?*
* if we start actually creating the pipelines around it, you will probably find that there are less steps as mentioned in the point above. With speculate I imagine we can also simplify this process further by automating the bumping of versions.

*The alternative approach*
The alternative approach you are suggesting is removing the Npm wrapping of the api. If we didn’t have the wrapper, we then would add a few things to the api repo, such as infrastructure and would add speculate to configure the api, which is also very bbc specific. At this point, closing off the repo maybe desirable, but it’s beneficial to be able to see what integration would look like.

*tl;dr*
* All in all what we gain from npm is the abstraction and reusability.
• I think there is little overhead for us in packaging if we set up the CI correctly.
• In terms of developing something useful for others in the NPM as a stand-alone API, this is very different.  This is what might be strange and not a typical NPM package. The reason why we are splitting out the API and the client initially is the security concern. If there were none, we would have allowed a single express server to handle both API and client. For external people, the API is designed to be used along side the client locally and needs to be usable without the concern of RPM management and AWS. Having a consistent layout (packaged as NPM) allows for easier thinking when considering the alternative back ends - CEP and Electron. We don’t want the infrastructure when packaging client, API and CEP together or client, API and Electron. (edited) 

What an essay !
I know some parts may have not made sense, we can chat over this tomorrow @james :3

---
James Dooley [11:35 AM]
Why have CI for npm release?
> - ensure clean environment every deployment and clear dependency - every computer that we have has a different local department set up so we want to make sure we have a clear idea on dependency.

This isn't a huge bone of contention but I do think the case you were making for it wasn't _totally_ necessary. Following good git behaviour - committing to master with PRs, creating releases etc. NPM publish is looking at GitHub master, the latest release and package.json, so that shouldn't affect things.

> - This is also another reason why we want a sandbox generally, to be able to replicate what would happen on Jenkins or a live cosmos machine (both centOS). It also provides transparency for other team members to see if test have succeeded or not.

> - automate testing

Sure. But technically if tests aren't passing, we shouldn't be pushing to master. Is automating testing any different to something like `"npm run publish": "npm test && npm publish"`? If we are to automate the versioning what's to say things don't get out of hand quickly - having to jump between 4-8 repos all being prototyped quickly and out in the open?

> I’m sure there are other reasons but I really think we should have CI and not do just local development and npm release even if we have a good practice of cleaning up before committing to master. There have been times when locally tests worked, the tests then failed in Jenkins due to some random dependency. This isn’t necessarily because the tests were poorly written or wasn’t exhaustive enough.

This doesn't tend to happen in my experience - the thing we need to maybe worry about here is the fact that we develop on completely different versions of Node and NPM to what's available on Jenkins.
There is overhead of having the API as an npm.

> - The alternative approach you are suggesting is removing the Npm wrapping of the api. If we didn’t have the wrapper, we then would add a few things to the api repo, such as infrastructure and would add speculate to configure the api, which is also very bbc specific. At this point, closing off the repo maybe desirable, but it’s beneficial to be able to see what integration would look like.

It just doesn't make a lot of sense to publish the server. Specific headers, configurations, CORS, middleware etc. are for people to decide on and deploy for their use. The behaviour of this API is not actually reusable for anyone else in any purpose despite saying this. We have ultra specific config for handing stuff to S3 buckets, SNS topics & queues, postgresQL database tables etc. Even if this is parameterised (...it won't be), we have talked about standing these APIs up with test & live environments ourselves too.

Speculate is a one liner in an npm script - and is open source. It doesn't make assumptions that you use Cosmos and is purely to generate Spec files for node projects - it doesn't configure the API as such - It's a useful abstraction here.

There's also the fact that external people won't even be using the same setup or resources. We have tried to abstract the infrastructure so people can spin up on Reith / GCP etc. etc. but then what use is the API anyway? You also have said that everything goes through localhost instead of the internet. What's the point of this for other people who would have to couple the two via CI & Jenkins and Spec files etc. themselves also?

> All in all what we gain from npm is the abstraction and reusability. I think there is little overhead for us in packaging if we set up the CI correctly.

Abstraction yes, reusability I don't think so. The overhead is having to manage Spec files, makefiles and system.d config, constantly reversioning and rebuilding, the CI pipeline to basically run an `npm publish` command, and having to change all of these so often whenever something changes. As well as the fact that in it's current state, _a lot is going to change_. How do you manage the versioning? As one version bumps, the other has to also?

Additionally, the infrastructure repo now has a bunch of code and it's own package.json node projects which pull in things and have to be maintained etc. It's no longer just infrastructure..?

> The reason why we are splitting out the API and the client initially is the security concern. If there were none, we would have allowed a single express server to handle both API and client.

These don't need to be on the same box at all, nor was I saying they should be. What is the security concern that having everything in one RPM addresses?

> For external people, the API is designed to be used along side the client locally and needs to be usable without the concern of RPM management and AWS.

How would this even work locally for someone else? Just serving dummy files? (edited) 
:face_with_rolling_eyes: _urgh_ i'm sorry for this wall of text
I also don't think i'm really capturing the points I was making yesterday
but I do think some of the points Eimi has made are wishful thinking / or not actually explained

**Comment by emettely** [comment]
> Do we need to revise the AWS architecture Diagram as well?
> Am I right in saying that in the latest setup
> 
>     * Client and server are sitting onto 2 distinct EC2 instances
> 
>     * and client one is the only public facing that does certificate based user authentication
> 
>     * client re-routes request to API server
> 
>     * API server can only talk to 

I checked the other ADR (https://github.com/bbc/digital-paper-edit-client/blob/master/docs/ADR/2019-04-23-transcript-architecture.md) with the architectural diagram and it's good as it is. We don't need to update it, because the single EC2 instance was part of this ADR. What it doesn't mention is the S3 signed URL stuff, but that's fine because that's the implementation detail. It'll be recorded in the README once we start working on the API.

**Comment by emettely** [comment]
#30 Adr
The right is original diagram (Eimi)
The left is proposed diagram (James)
The green marked bits are things we disagreed on.
We agreed to do the left one.
